### PR TITLE
Added method to wait until sleep function is in reality sleeping

### PR DESCRIPTION
**Pull request by sefirosweb**
- [x] The [FAQ](https://github.com/PrismarineJS/mineflayer/blob/master/docs/FAQ.md) doesn't contain a resolution to my issue 

## Versions
 - mineflayer: 4.3.0
 - server: vanilla 1.18.1
 - node: Dockerized:  node:16.13-bullseye

## Detailed description of a problem
If the bot click on bed at same time when other bot/player click on bed and steal him they don't return any error,
The function no return an error

I attach video you can see a bit better: https://youtu.be/S5PrYqO4KRk

## Expected behavior
Throw an error

### CODE REVIEW

**Comment by sefirosweb** [comment]
How i can rerun tests?

**Comment by amoraschi** [comment]
Push some changes, but doesn't the sleeping duration vary with the server's TPS?

**Comment by sefirosweb** [comment]
> Push some changes, but doesn't the sleeping duration vary with the server's TPS?

Yes, but the idea is not to get the sleeping duration, only they detect if bot are sleeping or not,
Because they don't return if the bot can be clicked correctly to bed

**Comment by sefirosweb** [comment]
Seems this PR break the test of sleep, checking..

**Comment by sefirosweb** [comment]
@amoraschi done!

I think now is much better, (See code)

the sleep function waits until the sleep event is triggered,

If there is no error and the event is not triggered then it returns an error indicating that it has not been able to sleep
So in this way we will avoid errors that the .catch event of sleep() is never fired

I removed in the test "await once(bot, 'sleep')" because sleep now wait to this trigger

File: lib/plugins/bed.js [hunk 1]
```diff
@@ -147,15 +147,26 @@ function inject (bot) {
...
-  bot._client.on('game_state_change', (packet) => {
-    if (packet.reason === 0) {
-      // occurs when you can't spawn in your bed and your spawn point gets reset
-      bot.emit('spawnReset')
-    }
-  })
+  async function waitUntilSleep () {
+    return new Promise((resolve, reject) => {
+      const timeoutForSleep = setTimeout(() => {
+        resolve('bot is not sleeping')
```
Comment (amoraschi) [inline lib/plugins/bed.js:161]:
```suggestion
        reject('bot is not sleeping')
```
Shouldn't this be `reject`?

**Review by amoraschi — CHANGES_REQUESTED** [review]
Promise should reject instead of resolve inside the `setTimeout`

File: lib/plugins/bed.js (hunk 1 above)
Comment (sefirosweb) [reply lib/plugins/bed.js:161]: I tried, but with reject don't received data,
Also they don't throw an error correctly

Comment (sefirosweb) [reply lib/plugins/bed.js:161]: With reject
image

Comment (U5B) [reply lib/plugins/bed.js:161]: try `reject(new Error('bot is not sleeping'))`

Comment (sefirosweb) [reply lib/plugins/bed.js]: @U5B you are great i learned a new thing :D

**Review by u9g — APPROVED** [review]

**Comment by u9g** [comment]
Thanks

**Pull request merged**